### DISCUSSION TITLE
Are macros with "inline matchers" possible?

### DISCUSSION THREAD

**Discussion by Elias-Graf** [lalrpop/lalrpop#972] [Q&A]
Let's suppose I have the following matcher:

```rust
pub DefVar: TopLevelExpr = {
    <left: @L> "("  "defvar"  ")" <right: @R> => {
        errors.push(ParseError::ExpectedDefVarName {
            err_span: (left, right).into(),
        });
        TopLevelExpr::Err
    },
    <left: @L> "(" "defvar" Symbol ")" <right: @R> => {
        errors.push(ParseError::ExpectedDefVarValue {
            err_span: (left, right).into(),
        });
        TopLevelExpr::Err
    },
    "(" "defvar" <name:Symbol> <value:Atom> ")" => {
        DefVar::new(name, value).into()
    },
};
```

Given the nature of the syntax, pretty much every expression is surrounded by parenthesis. With my goal to provide as good of an error message as possible, I'm looking to make the parenthesis "optional", and push an error if that was the case. So I tried creating the following macro:

```rust
Paren<Rule>: TopLevelExpr = {
    "(" <Rule> ")" => <>,
    <l:@L> <rule:Rule> ")" <r:@R> => {
        errors.push(ParseError::ExpectedParenthesisOpen {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
    <l:@L> "(" <rule:Rule> => {
        errors.push(ParseError::ExpectedParenthesisClose {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
};
```
Now, I would basically like to inline my complete matcher in this macro, but I'm not sure how that would syntactically work, even given just `@L`, and `@R`, probably not like this:

```rust
pub DefVar: TopLevelExpr = <l:@L> Paren<{
    ... => ...,
    ... => ...,
    ... => ...,
}>  <r:@R>;
```

Maybe something like this would be possible?

```rust
pub DefVar: TopLevelExpr = {
    <left: @L> <Paren<("defvar")>> <right: @R> =>...,
    <left: @L> <Paren<("defvar" Symbol)>> <right: @R> => ...,
    <Paren<("defvar" <name:Symbol> <value:Atom>)>> => ...,
};
```

What currently seems to work, is creating an "Inner" parser that does not contain any parenthesis, e.g.:

```rust
DefVarInner: TopLevelExpr = {
    <left: @L> "defvar" <right: @R> => ...,
    ...
};
```

and then creating the normal parser like this:

```rust
pub DefVar = Paren<DefVarInner>;
```

But now, `left`, and `right` are inside the parenthesis, which isn't terrible, but not what I wanted. In addition, I would have to declare two of every parser, which is a bit "meh" as well.

Is there any clever trick / syntax I'm missing?

Thanks for any help and pointers in advance! :)

**Comment by dburgener** [comment]
Could you separate the error creation from populating the span?

```
grammar;

Spanned<Rule> = {
    <left: @L> Paren<Rule> <right: @R> => {
        Rule.span(left, right)
    }
};
```

Then don't use `@L` and `@R` at all in the other rules, and on your return types, define a `Span` trait with `fn span()`.  Then you can do the "inner" approach you mentioned at the end, but still get the location matching outside the parentheses.

> In addition, I would have to declare two of every parser, which is a bit "meh" as well.

I don't follow this part.

An alternative approach, might be to just make all the values options:

```
grammar;

DefVar: TopLevelExpr = {
    <l: @L> "("? "defvar" <name:Symbol>? <value:Atom>? ")"? <r: @R> = {
        // action code
    }
};
```

And then check each of the options for None in the action code and return errors accordingly.  I haven't thought too hard about whether this approach ends up equivalent after macro expansion or not though.  Possibly it would cause parse generation errors.

**Reply by Elias-Graf** [reply]
Thank you so much for your response!

About this:

```rs
Spanned<Rule> = {
    <left: @L> Paren<Rule> <right: @R> => {
        Rule.span(left, right)
    }
};
```

`Rule` is just another parser, right, so I would implement a trait for the parser... and then... uh, I don't think I can follow. What's the trait implementation? Just a call to `parse`? If yes, how would I get the location back into the parser / grammar?

I've actually implemented some of the parsers like this:
```rs
DefVar: TopLevelExpr = {
    <l: @L> "("? "defvar" <name:Symbol>? <value:Atom>? ")"? <r: @R> = {
        // action code
    }
};
```

But at some point I switched to doing all the parsing logic with the matchers. I could try and see if doing it like that works when all parsers do it.

I can try and implement the second approach and get back to you this evening, but I'd still be quite interested what you meant my the first solution.
And thanks again!

Oh, and:

> In addition, I would have to declare two of every parser, which is a bit "meh" as well.

Don't worry about it too much, I was just not too much of a fan of having an extra "Inner" parser for every parser, but it's really not that bad 😅.

**Reply by Elias-Graf** [reply]
I'm actually just wondering, does the macro respect and know about the "defvar" that is inside the "Rule"? Because if not, that will lead to conflicts with every other parser, once I apply it to them... I will have to try!

**Reply by dburgener** [reply]
> Rule is just another parser, right, so I would implement a trait for the parser... and then... uh, I don't think I can follow. What's the trait implementation? Just a call to parse? If yes, how would I get the location back into the parser / grammar?

Sorry, there are mistakes in the code snippet I wrote, which confused you.  Let me try again to fix those and be a little more explicit in the description.

```
grammar;

Spanned<Rule> = {
    <left: @L> <rule: Rule> <right: @R> => {
        rule.span(left, right)
    }
};

Paren<Rule> = {
    "(" DefVarInner ")"
    // etc
};

DefVarInner = {
    "defvar",
    // etc
};

DefVar = {
    Spanned<Paren<DefVarInner>>
}
```

and in your rust code:

```
pub trait Spanned {
    fn span(left: usize, right: usize);
}

impl Spanned for TopLevelExpr {
    fn span(&mut self, left: usize, right: usize) {
        match self {
              TopLevelExpr::Err(e) => {
                  self.span = Some((left, right)); // or however its stored
              }
              _ => ()
        }
    }
}
```

Hopefully that's more explicit and with fewer mistakes so the idea is more clear?

As you'll note from the above, a downside of this, is that spans become optional, since they aren't set at creation time, but shortly after.

**Reply by dburgener** [reply]
> I'm actually just wondering, does the macro respect and know about the "defvar" that is inside the "Rule"? Because if not, that will lead to conflicts with every other parser, once I apply it to them... I will have to try!

I'm not sure I follow the question, but macros are expanded before the parser is generated, so they effectively inline and expand out to all the rules you would have written by hand.  Maybe that answers the question?

**Reply by Elias-Graf** [reply]
Yes, I think I understand now. But, I don't really need the span on the rule / the resulting AST node. As you can see in my original example:

```rs
<left: @L> "("  "defvar"  ")" <right: @R> => {
    errors.push(ParseError::ExpectedDefVarName {
        err_span: (left, right).into(),
    });
    TopLevelExpr::Err
}
```

...I push an error into a vector and just return a simply dummy AST node `Err`. Thinking about it, I could maybe just have the error inside the `Err` variant of my top level expression. I was simply following [the book](https://lalrpop.github.io/lalrpop/tutorial/008_error_recovery.html), so I'm not really sure of any (dis)advantages...
What you said about the consequence of it being optional isn't great, though :(. Errors really should always have a span, and I would like to leverage rust's type system.

Maybe I can somehow pass the state from the `Paren` passer to the inner rule? I know from [this chapter](https://lalrpop.github.io/lalrpop/tutorial/009_state_parameter.html), and from my errors, that it's possible to pass a "global" state. But in this case, I would need it only for some of the parsers, and very local for every single pass...

It would be really neat, if I could write something like this:

```rs
pub DefVar: TopLevelExpr = <l:@L> Paren<{
    "defvar" =>...,
    "defvar" Symbol  => ...,
    "defvar" <name:Symbol> <value:Atom> => ...,
}>  <r:@R>;
```

_(yes I know this syntax doesn't make sense, but I hope it gets the idea across somewhat)_

And it would generate my all the possible options. But I'm assuming that's not possible, as you would probably have told me about it :D.

> I'm not sure I follow the question, but macros are expanded before the parser is generated, so they effectively inline and expand out to all the rules you would have written by hand. Maybe that answers the question?

Yeah, that does answer my question. Was worried I'd get a state, where it would look for _n_ "(", followed by a "black box" rule.

The easiest, and most practical thing might be, to just live with the fact that my error spans aren't going to capture the complete "unit" of an "object" / "definition"? Do you have a recommendation, maybe?

PS:
What I had suggested earlier, would probably make way more sense syntactically and structurally, somethink like:

```rs
pub DefVar: TopLevelExpr = {
    <left: @L> Paren<( "defvar" )> <right: @R> =>...,
    <left: @L> Paren<( "defvar" Symbol )> <right: @R> => ...,
    Paren<( "defvar" <name:Symbol> <value:Atom> )> => ...,
};
```

~~Somehow the "inner rule", inside parenthesis, would have to be passed to the `Paren` parser though, with context (`@L`, `@R`) from outside... hmmm, tricky. Sounds vaguely similar to a closure / lambda, though.~~ Wait, no, if it's expanded like you said ealier, this might actually be possible, no?

**Reply by dburgener** [reply]
> ...I push an error into a vector and just return a simply dummy AST node Err. Thinking about it, I could maybe just have the error inside the Err variant of my top level expression. I was simply following [the book](https://lalrpop.github.io/lalrpop/tutorial/008_error_recovery.html), so I'm not really sure of any (dis)advantages...
> What you said about the consequence of it being optional isn't great, though :(. Errors really should always have a span, and I would like to leverage rust's type system.

Oh yes, I had forgot that aspect.  You could still do the same approach though.  Just expand your `TopLevelExpr::Err` type to include the message, and then add the span in and create a ParseError from the top level `Spanned` macro.

> What you said about the consequence of it being optional isn't great, though :(. Errors really should always have a span, and I would like to leverage rust's type system.

Yes, not great.  Although the above actually sidesteps that I think.  In general, you could always create an intermediary type with the `Optional` and then transform that into one without optional, returning an internal error of some sort on None to get the type safety later.  But like I said, if you do the approach in the above paragraph, I don't think you need `Optional` at all.

> Maybe I can somehow pass the state from the Paren passer to the inner rule? I know from [this chapter](https://lalrpop.github.io/lalrpop/tutorial/009_state_parameter.html), and from my errors, that it's possible to pass a "global" state. But in this case, I would need it only for some of the parsers, and very local for every single pass...

If we have support for something like this today, I'm not aware of it.

> And it would generate my all the possible options. But I'm assuming that's not possible, as you would probably have told me about it :D.

I actually think that approach might be possible.  [This test](https://github.com/lalrpop/lalrpop/blob/b8b82e9ecde5ca531efd13ce1beb9c1c5fa488a1/lalrpop/src/normalize/macro_expand/test.rs#L136-L137) seems to do essentially that.  Try it and see?  I don't think this is behavior that we test/advertise much, but it might work today, and if it doesn't, I'd be interested in at least looking into what it would take to make it work.

> PS:
> What I had suggested earlier, would probably make way more sense syntactically and structurally, somethink like:

(snipped)

Yeah, that seems reasonable to me.  I would definitely try that and report back if it doesn't work as expected.

**Reply by Elias-Graf** [reply]
Thank you for your patience!

I've run into a few of issues. First, I'd like to simplify my example a bit and reduce dependencies. So I've replaced my `Atom`, and `Symbol` parsers with `"stringLiteral"`, which is the name assigned to string literal tokens in my custom lexer. I have kept `TopLevelExpr`, and my errors (`ParseError`), since they are relative simple rust enums, and shouldn't change anything about the parser.

Now, let's look at the follow two parsers:

```rs
pub Test: TopLevelExpr = {
    Paren<("defvar")> => todo!(),
}

Paren<I>: TopLevelExpr = {
    "(" <I> ")" => <>,
    <l:@L> <rule:I> ")" <r:@R> => {
        errors.push(ParseError::ExpectedParenthesisOpen {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
    <l:@L> "(" <rule:I> <r:@R> => {
        errors.push(ParseError::ExpectedParenthesisClose {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
    <I> => todo!(),
};
```

This runs into the error:

```rs
      --> /home/elias/Git/eww-yuck-lsp/target/debug/build/potato-aa6bcb80d5fbd083/out/grammar.rs:227599:5
       |
227597 | ) -> TopLevelExpr
       |      ------------ expected `TopLevelExpr` because of return type
227598 | {
227599 |     __0
       |     ^^^ expected `TopLevelExpr`, found `Tok<'_>`
```

I found the cause for that. It seems to be the `"(" <I> ")" => <>,` matcher in the `Paren` parser. Thinking about it, I don't really need that rule anyway, because in the success case, I would like to handle the values in the `Test` parser anyway, so let's remove that rule for now.

With that removed, I can successfully compile the project, so let's move on to the success case, where I would like to construct the AST node. The following code should be sufficent for a "does that compile?" test:

```rs
pub Test: TopLevelExpr = {
    <l:@L> Paren<("defvar" <n:"literalString"> <v:"literalString">)> <r:@R> => todo!(),
}
```

This now leads to:

```
  process didn't exit successfully: `/home/elias/Git/eww-yuck-lsp/target/debug/build/potato-f9ab5a36758aa54e/build-script-build` (exit status: 101)
  --- stdout
  processing file `/home/elias/Git/eww-yuck-lsp/lib/potato/src/grammar.lalrpop`
  /home/elias/Git/eww-yuck-lsp/lib/potato/src/grammar.lalrpop:92:18: 92:67 error: named symbols like `n:"literalString"` are only allowed at the top-level of a nonterminal

  --- stderr
        <l:@L> Paren<("defvar" <n:"literalString"> <v:"literalString">)> <r:@R> => todo!(),
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  thread 'main' panicked at lib/potato/build.rs:2:57:
  called `Result::unwrap()` on an `Err` value: Kind(InvalidData)
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Removing the named symbols results in successful compilation. That isn't all that useful (I think), but let's continue experimenting, and implement both my error cases:

```rs
pub Test: TopLevelExpr = {
    <l:@L> Paren<( "defvar" )> <r:@R> => {
        errors.push(ParseError::ExpectedDefVarName {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
    <l:@L> Paren<( "defvar" "literalString" )> <r:@R> => {
        errors.push(ParseError::ExpectedDefVarValue {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
};
```

And it in fact compiles! BUT, it doesn't produce any useable result, since I needed to remove the "success" case from the `Paren` parser earlier (`"(" <I> ")" => <>`), and so I get an: "UnrecognizedToken: DelimiterRightParen", since nothing matches `"(defvar)"`.

**Reply by Elias-Graf** [reply]
Btw. I took the time to extract the parser from the rest of my thingy, and push it to a repository: https://github.com/Elias-Graf/rotten-potato. In case that helps with anything :)

**Reply by dburgener** [reply]
> And it in fact compiles! BUT, it doesn't produce any useable result, since I needed to remove the "success" case from the Paren parser earlier ("(" <I> ")" => <>), and so I get an: "UnrecognizedToken: DelimiterRightParen", since nothing matches "(defvar)".

I'm not sure I followed all the steps in here correctly, but if the issue is that everything works except you had to remove the`"(" <I> ")"` case, that's very fixable.  You were just returning `I`, whatever type that happens to be.  You need to in the action code build a `TopLevelExpr` from `I` and return that so that the return type from `Paren<_>` is consistent.

**Reply by Elias-Graf** [reply]
Yes, but it seems like `I` is not actually the result of the parser, but `Tok<_>` instead. I sadly don't understand enough of the internals, to deduce why exactly this is happening. If the issue was simply, that parsers may return different things, I'd just simply return `<>.into()`b (and of course implement it for all the different structs), but as far as I can tell from the error message, it's not a consumed value at all, no?
Maybe I just didn't understand what you were saying, though.

And I don't really understand, how that would work. The action code is this... right? 

```rust
    Paren<( "defvar" <name:Symbol> <value:Atom> )> => ...,
```

**Reply by dburgener** [reply]
So `Tok<_>` is the *type* of `I`, which is the token enum for your custom lexer.  The *value* of `I` will be whatever you passed to `Paren` in the places you used the macro.  So `"defvar"` for example.

The action code is the part in the `{` `}` brackets after the grammar line.  That's the part of the code that you write for what to do when the parser hits a particular rule.

**Reply by Elias-Graf** [reply]
Yes, I know. Let me try to go top down again. Assuming the following code:

```rust
pub DefListenTest: TopLevelExpr = {
    Paren<( "defvar" <name:Symbol> <value:Atom> )> => TopLevelExpr::Err,
};

Paren<I>: TopLevelExpr = {
    "(" <I> ")" => <>,
    <l:@L> <rule:I> ")" <r:@R> => {
        errors.push(ParseError::ExpectedParenthesisOpen {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
    <l:@L> "(" <rule:I> <r:@R> => {
        errors.push(ParseError::ExpectedParenthesisClose {
            err_span: (l, r).into(),
        });
        TopLevelExpr::Err
    },
};
```

I get:

```
Caused by:
  process didn't exit successfully: `/home/elias/Git/rotten-potato/target/debug/build/rotten-potato-68adc91569ff64c7/build-script-build` (exit status: 101)
  --- stdout
  processing file `/home/elias/Git/rotten-potato/src/grammar.lalrpop`
  /home/elias/Git/rotten-potato/src/grammar.lalrpop:49:11: 49:49 error: named symbols like `name:Symbol` are only allowed at the top-level of a nonterminal

  --- stderr
        Paren<( "defvar" <name:Symbol> <value:Atom> )> => TopLevelExpr::Err,
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  thread 'main' panicked at build.rs:2:57:
  called `Result::unwrap()` on an `Err` value: Kind(InvalidData)
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Removing the named parameters ***which is of no use to me (I think), since now I can't construct a proper AST-node***, but I'll remove them and continue to display further issues:


```rust
pub DefListenTest: TopLevelExpr = {
    Paren<( "defvar" <Symbol> <Atom> )> => TopLevelExpr::Err,
};
```

results in even more issues:

```
    Blocking waiting for file lock on build directory
   Compiling rotten-potato v0.1.0 (/home/elias/Git/rotten-potato)
error[E0308]: mismatched types
      --> /home/elias/Git/rotten-potato/target/debug/build/rotten-potato-36b8e843c964e903/out/grammar.rs:223630:5
       |
223628 | ) -> TopLevelExpr
       |      ------------ expected `TopLevelExpr` because of return type
223629 | {
223630 |     __0
       |     ^^^ expected `TopLevelExpr`, found `(Symbol, Atom)`
       |
       = note: expected enum `TopLevelExpr`
                 found tuple `(Symbol, Atom)`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `rotten-potato` (lib) due to 1 previous error
```

The action code *does* in fact return a `TopLevelExpr`, in this case, just a hardcoded `Err`.

**Reply by Elias-Graf** [reply]
`I` does not seem to be the result of the action code, but the extracted parameters.

**Reply by dburgener** [reply]
> `I` does not seem to be the result of the action code, but the extracted parameters.

Yes, that's exactly what I said in my previous reply.  The issue with the `"(" <I> ")" => <>,` part is that you are returning `I`, which is either a `Tok<_>`, or a tuple of `Tok`s, depending on the macro call.  It's not a TopLevelExpr.  You need to build one in the action code for that rule.

However, it does seem like your real problem here is the error about named symbols not at the top level of a non-terminal.  That brings me back to this suggestion I made earlier:

> Oh yes, I had forgot that aspect. You could still do the same approach though. Just expand your TopLevelExpr::Err type to include the message, and then add the span in and create a ParseError from the top level Spanned macro.

Is there some reason you can't do that in order to avoid passing the tuples into the macro?  I don't know of a clean way to get around the top level non-terminal error in this case, and the tuple usage complicates things.

**Reply by Elias-Graf** [reply]
> Yes, that's exactly what I said in my previous reply. The issue with the "(" <I> ")" => <>, part is that you are returning I, which is either a Tok<_>, or a tuple of Toks, depending on the macro call. It's not a TopLevelExpr. You need to build one in the action code for that rule.

Could you show me how that would work? I really don't understand this part. Optimally, I could call the action code defined in `DefVarTest` / `DefListenTest` (I've used `DefListenTest` by mistake earlier, I'll be using `DefVarTest` for future examples, since it's quite a lot simpler):

```rust
pub DefVarTest: TopLevelExpr = {
    Paren<("defvar" <name:Symbol> <value:Atom>)> => {
        DefVar::new(name, value).into()
    },
};
```

> Oh yes, I had forgot that aspect. You could still do the same approach though. Just expand your TopLevelExpr::Err type to include the message, and then add the span in and create a ParseError from the top level Spanned macro.

Yes, I could certainly include the message in the AST, but I fail to see why this would benefit me here. I used `TopLevelExpr::Err` as a simple placeholder, normally I would use:

```rust
pub DefVarTest: TopLevelExpr = {
    Paren<("defvar" <name:Symbol> <value:Atom>)> => {
        DefVar::new(name, value).into()
    },
};
```

> Is there some reason you can't do that in order to avoid passing the tuples into the macro? I don't know of a clean way to get around the top level non-terminal error in this case, and the tuple usage complicates things.

I'm not quite sure what tuple you're talking about? The one resulting from the fact I have two parameters? I *mean*, I could try and get around that with:

```rust
pub DefVarTest: TopLevelExpr = {
    Paren<("defvar" <Symbol> <Atom>)> => {
        let (name, value) = <>;
        DefVar::new(name, value).into()
    },
};
```

But it really seems like the system is not designed to work in this way. In the `Paren` macro, I receive the parameters, and now with this code, I'm noticing that I receive the `TopLevelExpr`, probably from `Paren` instead:

```
error[E0308]: mismatched types
      --> /home/elias/Git/rotten-potato/target/debug/build/rotten-potato-36b8e843c964e903/out/grammar.rs:222232:13
       |
222232 |         let (name, value) = __0;
       |             ^^^^^^^^^^^^^   --- this expression has type `TopLevelExpr`
       |             |
       |             expected `TopLevelExpr`, found `(_, _)`
       |
       = note: expected enum `TopLevelExpr`
                 found tuple `(_, _)`
```
